C# Unit 2 Auto-graded Assignments (#133)
LaunchCodeEducation/csharp-web-development: src/assignments/tech-jobs-oo.rst

Your Assignment
---------------

Running the Auto-grading Tests
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Before diving into your tasks, review ``Assignment #0: Hello, World! -- NEED LINK`` for details on 

Author: ```suggestion
Before diving into your tasks, review :ref:`Assignment #0: Hello, World! <assignment0>` for details on 
```

running the auto-grading tests for this assignment. 

This assignment has many more tests than the previous two, so we’ve organized them 
into separate files. The tests are visible in the Project's Solution, but they won't run from there.

.. figure:: figures/Tests-In-Solution-A2OO.png
   :scale: 50%

>    :alt: List of projects in the solution.  Tests are in "TechJobsOOTest" project.

Author: ```suggestion
   :alt: List of projects in the solution.  Auto-grading tests are in "TechJobsOOTest" project.
```

Each of the one test files, ``TestTaskX``, contains the tests to grade one of the six tasks 
outlined in below (there’s no coding work for your first task, so there is **no** ``TestTask1``). 

In order to run any of the tests, we need to use Visual Studio's ``Test Explorer``.  
Open ``Test Explorer``, but do NOT ``Run All Tests`` at this time.  
Visual Studio will not be able to run many of these tests at this time because you have not 
any written code they are designed to test.  It will only throw errors.  

> If you did run ALL the test, stop the tests by clicking on the red square in ``Test Explorer``.

Author: ```suggestion
If you did run ALL the tests, stop the tests by clicking on the red square in ``Test Explorer``.
```

#. Generate the ``Equals()`` and ``GetHashCode()`` methods. Consider two ``Job``
   objects equal when their ``id`` fields match.

Run Your Tests
^^^^^^^^^^^^^^

Run your ``TestTask3`` unit tests.  Refactor your code as needed.  
Do not move onto Task 4 until you have passed all of Task 3's auto-grading unit tests.

.. admonition:: Tip

   Save, commit, and push your work to GitHub.

Task 4: Use Unit Testing to Verify Parts of the ``Job`` Class
--------------------------------------------------------------

.. admonition:: Note

>    This test has been commented out for the rest of the tests to run.

Author: Do you mean "task"? I am not entirely sure where this is.

Reviewer: yes, the entire class for the test.  i will ponder the wording on this.  it became an issue as i tried to reset it from an entire solution to starter code.  which was also right before i was leaving on vacation.  will think about this one and how to make it more clear.  ;D (and the other ones below)

that ``Equals()`` returns ``true`` when two objects have the same ID. However,
the positive test is irrelevant in this case.

The way you build your ``Job`` class, each ``id`` field gets assigned a unique
value, and the class does not contain a setter for the ``id`` field. You also verified
that each new object gets a different ID when you tested the constructors.
Without modifying the constructors or adding a setter, there is no scenario in
which two different jobs will have the same ID number. Thus, we can skip the
test for this condition.

Run Your Tests
^^^^^^^^^^^^^^

Run your ``TestTask4`` unit tests.  Refactor your code as needed.  
Do not move onto Task 5 until you have passed all of Task 4's auto-grading unit tests.

.. admonition:: Tip

   Time to save, commit, and push your work to GitHub again.

Task 5: Use TDD to Build The ``ToString()`` Method
--------------------------------------------------

.. admonition:: Note

   This test has been commented out for the rest of the tests to run.
   
   Remove the ``/*`` in **line 1** and ``*/`` in **line 167**.

> 

Author: Do you mean "task"? I am not entirely sure where this is.

#. Modify ``ToString()`` to make the new test pass. Also, make sure that your
   updates still pass all of the old tests.
#. Continue this test-refactor cycle until all of the behaviors we want for
   ``ToString()`` work. Remember to add only ONE new test at a time.

Test 3: ``TestToStringHandlesEmptyField``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#. Code your final test for the last required behavior, 
   then run the tests to make sure the new one fails. Call this test ``TestToStringHandlesEmptyField``.
#. Modify ``ToString()`` to make the new test pass.  Also make sure that your updates still pass all the old tests.

Cool! Your ``Job`` class is now complete and operates as desired.

Run Your Tests
^^^^^^^^^^^^^^

Run your ``TestTask5`` unit tests.  Refactor your code as needed.  
Do not move onto Task 6 until you have passed all of Task 5's auto-grading unit tests.

And don't forget to ``add``, ``commit`` and ``push`` the latest version of your code.

Task 6: Refactor to DRY the Support Classes
-------------------------------------------

.. admonition:: Note

   This test has been commented out for the rest of the tests to run.
   

>    Remove the ``/*`` in **line 1** and ``*/`` in **line 96**.

Author: I think you should clarify this one too! It took me a while to figure out where to un-comment

   ``PositionType`` classes.
#. Rerun your unit tests to verify that your classes and methods still work.

Run Your Tests
^^^^^^^^^^^^^^

Run your ``TestTask6`` unit tests.  Refactor your code as needed.  

> This is the last set of test!  Congrats!  

Author: ```suggestion
This is the last set of tests!  Congrats!  
```

--------------

To turn in your assignment and get credit, follow the

> ``SHOULD LINK BACK TO HELLO WORLD AUTOGRADER'S SUBMISSION INSTRUCTIONS``

Author: ```suggestion
:ref:`submission instructions <submitting-your-work>`.
```
